### PR TITLE
Add support for multiple composefile when deploying

### LINKED ISSUE #30127: Deploy to swarm from multiple compose files
I'm currently testing the new `--compose-file` parameter for `docker stack deploy` in 1.13.0-rc6, but my use case requires deploying from multiple compose files (as possible with `docker-compose -f base.yml -f additionalService.yml -f production.yml`). At the moment, using globs or multiple `--compose-file` is not supported as far as I can see. Are there plans to support this in future releases? 

A simple workaround/implementation strategy is to merge the compose files before deploying, but I would rather want to avoid using temporary files.

### PR DESCRIPTION
**Pull request by vdemeester**
Using a custom version of mergo to manage special cases. 

- [x] Fix vendor (right now it's a locally modified version of mergo)
- [x] <del>Handling `docker-compose.override.yml` per default</del>
- [x] More tests on edge cases compared to `docker-compose`

Should fix moby/moby#30127

/cc @shin- too for edge case in the merge that I would have missed

*as of now, it is supposed to be red (i.e. vendor should not validate)*

boom

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

### CODE REVIEW

File: cli/command/stack/deploy.go [hunk 1]
```diff
@@ -63,9 +63,9 @@ func runDeploy(dockerCli command.Cli, opts deployOptions) error {
 	}
 
 	switch {
-	case opts.bundlefile == "" && opts.composefile == "":
+	case opts.bundlefile == "" && len(opts.composefile) == 0:
```
Comment (ksouf) [inline cli/command/stack/deploy.go]: i think we should keep the symmetry :)
i mean on checking the emptiness of a string always using len function instead of '== "" '

File: cli/command/stack/deploy.go [hunk 2]
```diff
@@ -63,9 +63,9 @@ func runDeploy(dockerCli command.Cli, opts deployOptions) error {
...
 		return errors.Errorf("Please specify either a bundle file (with --bundle-file) or a Compose file (with --compose-file).")
-	case opts.bundlefile != "" && opts.composefile != "":
+	case opts.bundlefile != "" && len(opts.composefile) != 0:
```
Comment (ksouf) [inline cli/command/stack/deploy.go]: same thing :)

File: cli/compose/loader/loader_multiple_test.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+			}},
+			{Filename: "override.yml", Config: map[string]interface{}{
+				"version": "3.0",
+			}},
+		},
+	}
+	_, err := Load(configDetails)
+	require.Error(t, err)
+}
+*/
```
Comment (ksouf) [inline cli/compose/loader/loader_multiple_test.go]: i think we should remove commented code :)

File: cli/compose/loader/loader_multiple_test.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+				"cap_add": []interface{}{
+					"NET_ADMIN",
+				},
+			},
+		},
+		"volumes":  map[string]interface{}{},
+		"networks": map[string]interface{}{},
+		"secrets":  map[string]interface{}{},
+		"configs":  map[string]interface{}{},
+	}
```
Comment (ksouf) [inline cli/compose/loader/loader_multiple_test.go]: mmmm can we extract something ? :)

File: cli/compose/loader/loader_multiple_test.go [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+			},
+			"bar": map[string]interface{}{
+				"image": "bar",
+			},
+		},
+		"volumes":  map[string]interface{}{},
+		"networks": map[string]interface{}{},
+		"secrets":  map[string]interface{}{},
+		"configs":  map[string]interface{}{},
+	}
```
Comment (ksouf) [inline cli/compose/loader/loader_multiple_test.go]: same here i think we can extract it

File: cli/compose/loader/loader_multiple_test.go (hunk 3 above)
Comment (vdemeester) [reply cli/compose/loader/loader_multiple_test.go]: @ksouf oh right, I kept this one to discuss it with @dnephin and @shin- if we do support merging 2 different version on `docker-compose` or not.

File: cli/compose/loader/loader_multiple_test.go (hunk 4 above)
Comment (vdemeester) [reply cli/compose/loader/loader_multiple_test.go]: I'm going to write a builder for that once it's in `code-review` yes 😉

File: cli/compose/loader/loader_multiple_test.go (hunk 3 above)
Comment (ksouf) [reply cli/compose/loader/loader_multiple_test.go]: @vdemeester all right :)

Comment (vdemeester) [reply cli/compose/loader/loader_multiple_test.go]: oh but it doesn't really make sense in the PR because it's not the one implementing parsing of v1/v2 compose files 😅

Comment (dnephin) [reply cli/compose/loader/loader_multiple_test.go]: It is not supported, it's an error to combine different versions.

File: cli/command/stack/deploy.go [hunk 6]
```diff
@@ -22,7 +22,7 @@ const (
 
 type deployOptions struct {
 	bundlefile       string
-	composefile      string
+	composefile      []string
```
Comment (silvin-lubecki) [inline cli/command/stack/deploy.go]: Rename to composefiles ?

File: cli/compose/loader/loader.go [hunk 7]
```diff
@@ -45,55 +45,59 @@ func Load(configDetails types.ConfigDetails) (*types.Config, error) {
...
 
-	if services, ok := configDict["services"]; ok {
-		if servicesDict, ok := services.(map[string]interface{}); ok {
-			forbidden := getProperties(servicesDict, types.ForbiddenProperties)
+	for _, file := range configDetails.ConfigFiles {
+		configDict := file.Config
+
+		if services, ok := configDict["services"]; ok {
+			if servicesDict, ok := services.(map[string]interface{}); ok {
+				forbidden := getProperties(servicesDict, types.ForbiddenProperties)
```
Comment (silvin-lubecki) [inline cli/compose/loader/loader.go]:
```go
if forbidden := getProperties(servicesDict, types.ForbiddenProperties); len(forbidden) > 0{
    return nil, &ForbiddenPropertiesError{Properties: forbidden}
}
```

File: cli/compose/loader/loader.go [hunk 8]
```diff
@@ -45,55 +45,59 @@ func Load(configDetails types.ConfigDetails) (*types.Config, error) {
...
 
-	cfg.Secrets, err = LoadSecrets(config["secrets"], configDetails.WorkingDir)
-	if err != nil {
-		return nil, err
+		cfg.Secrets, err = LoadSecrets(config["secrets"], configDetails.WorkingDir)
+		if err != nil {
+			return nil, err
+		}
+
+		cfg.Configs, err = LoadConfigObjs(config["configs"], configDetails.WorkingDir)
```
Comment (silvin-lubecki) [inline cli/compose/loader/loader.go]: Why don't you check err here?

File: cli/compose/loader/merge.go [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+	"github.com/imdario/mergo"
+	"github.com/pkg/errors"
+)
+
+type specials struct {
+	m map[reflect.Type]func(dst, src reflect.Value) error
+}
+
+func (s *specials) Transformer(t reflect.Type) func(dst, src reflect.Value) error {
+	if fn, ok := s.m[t]; ok {
```
Comment (silvin-lubecki) [inline cli/compose/loader/merge.go]: Why don't you just return s.m[t]?
```go
func (s *specials) Transformer(t reflect.Type) func(dst, src reflect.Value) error {
    return s.m[t]
}
```
Isn't it equivalent?

File: cli/compose/loader/merge.go [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+		return fn
+	}
+	return nil
+}
+
+func merge(configs []*types.Config) (*types.Config, error) {
+	base := configs[0]
```
Comment (silvin-lubecki) [inline cli/compose/loader/merge.go]: Check configs length before access to the first element?

**Review by silvin-lubecki — COMMENTED** [review]
LGTM

File: cli/compose/loader/loader.go (hunk 8 above)
Comment (vdemeester) [reply cli/compose/loader/loader.go]: Oups, I forgot :sweat:

File: cli/compose/loader/merge.go (hunk 9 above)
Comment (vdemeester) [reply cli/compose/loader/merge.go]: It would be :+1:

File: cli/compose/loader/merge.go (hunk 10 above)
Comment (vdemeester) [reply cli/compose/loader/merge.go]: At this point, it's guaranteed that you have a least one config (and as it's not exported.. I think we can "not check" but open to change that)

**Comment by dnephin** [comment]
Hmm, jenkins build didn't trigger (or didn't update the github PR) for some reason. I've seen that happen occasionally.

> Handling docker-compose.override.yml per default

I don't think we should read this by default. This is similar to the `.env` file support, which is specific to the tool `docker-compose` not the format (which is shared).

> More tests on edge cases compared to docker-compose

Haven't looked at existing tests yet, but something that we need to make sure is tested is the case where we don't actually merge structs but have one struct override the other (ex: build).

File: cli/command/stack/deploy.go (hunk 1 above)
Comment (dnephin) [reply cli/command/stack/deploy.go]: I really dislike checking empty string using `len(string) == 0`. Sure it works, but it's a lot less obvious when reading the code. I think most people say "empty string", not "a string of zero characters", so `variable == ""` matches this phrase.

I'm in favour of keeping it as it is now.

**Review by dnephin — COMMENTED** [review]
Design LGTM

Finding all the necessary test cases will be difficult. There are a ton in docker-compose. If we can find all the places that don't fully merge structs that would be a good thing to test.

**Comment by thaJeztah** [comment]
Design LGTM as well; want to give this a bit of a try; ideally we'd also have a way to print the merged stack before deploying (similar to `docker-compose config`, could be, e.g. `docker stack print -c file -c file -c file`, so that I can check what will be deployed and if the merged results are what I expect them to be.

File: cli/compose/loader/loader.go [hunk 11]
```diff
@@ -45,27 +45,36 @@ func Load(configDetails types.ConfigDetails) (*types.Config, error) {
...
-	configDict := getConfigDict(configDetails)
-	configDetails.Version = schema.Version(configDict)
+	configs := []*types.Config{}
 
-	if err := validateForbidden(configDict); err != nil {
-		return nil, err
-	}
+	for _, file := range configDetails.ConfigFiles {
+		configDict := file.Config
+		configDetails.Version = schema.Version(configDict)
```
Comment (dnephin) [inline cli/compose/loader/loader.go]: I think this should only set the version from the first config, and error if any other config does not match that version.

File: cli/compose/loader/merge.go [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+			reflect.TypeOf(types.BuildConfig{}): func(dst, src reflect.Value) error {
+				dst.Set(src)
+				return nil
+			},
+		},
+	}
+	for name, overrideService := range overrideServices {
+		if baseService, ok := baseServices[name]; ok {
+			if err := mergo.Merge(&baseService, &overrideService, mergo.WithOverride, mergo.WithTransformers(specials)); err != nil {
+				// if err := mergo.MergeWithOverwriteAndTransform(&baseService, &overrideService, specials); err != nil {
```
Comment (dnephin) [inline cli/compose/loader/merge.go]: commented code

File: cli/compose/loader/merge.go [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+func merge(configs []*types.Config) (*types.Config, error) {
+	base := configs[0]
+	for _, override := range configs[1:] {
+		var err error
+		// merge services
+		base.Services, err = mergeServices(base.Services, override.Services)
+		if err != nil {
+			return base, errors.Wrap(err, "cannot merge services")
```
Comment (dnephin) [inline cli/compose/loader/merge.go]: Errors here might be different to debug when there are many files because there is no mention of which file failed to merge. Maybe `types.Config` needs a `Filename` field so that it can be included here?

File: cli/compose/loader/merge.go [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+				dst.Set(src)
+				return nil
+			},
+		},
+	}
+	for name, overrideService := range overrideServices {
+		if baseService, ok := baseServices[name]; ok {
+			if err := mergo.Merge(&baseService, &overrideService, mergo.WithOverride, mergo.WithTransformers(specials)); err != nil {
+				// if err := mergo.MergeWithOverwriteAndTransform(&baseService, &overrideService, specials); err != nil {
+				return base, errors.Wrap(err, "cannot merge compose file")
```
Comment (dnephin) [inline cli/compose/loader/merge.go]: This message seems more appropriate as a wrap around the error returned by `merge()`. At this level I think the detail that needs to be included is the service name.

File: cli/compose/loader/merge.go [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+	}
+	return base, nil
+}
+
+func mergeServices(base, override []types.ServiceConfig) ([]types.ServiceConfig, error) {
+	baseServices := mapByName(base)
+	overrideServices := mapByName(override)
+	specials := &specials{
+		m: map[reflect.Type]func(dst, src reflect.Value) error{
+			reflect.TypeOf(types.BuildConfig{}): func(dst, src reflect.Value) error {
```
Comment (dnephin) [inline cli/compose/loader/merge.go]: I thought that build was a non-recursive merge, but looking at the docker-compose code again it seems like maybe it does recursive?

`logging` looks like a special case, where it changes if the drier changes. I think driver options are a case where they don't get merged but overridden. I don't have a full list but I think I remember there being a bunch of them.

**Comment by imdario** [comment]
Sorry, I didn't see this before. I'll check it by this weekend.

**Comment by vdemeester** [comment]
Rebase and updated.. I need to take your comments into account @dnephin 😛

File: cli/compose/loader/merge.go (hunk 15 above)
Comment (vdemeester) [reply cli/compose/loader/merge.go]: hum 🤔 what should we do here ? 😛 
`logging` is a special case but build isn't ?
cc @shin-

Comment (shin-) [reply cli/compose/loader/merge.go]: Regarding `logging`, the idea is that options are driver-specific, thus an override of the `driver` resets the `options` dictionary. Unless the base driver is null or unset, in which case it's assumed the user is actively attempting to use inheritance, and base options are merged with the override. Same thing if the base has specified a driver but not the override.

These tests codify the behavior on the docker-compose side: https://github.com/docker/compose/blob/master/tests/unit/config/config_test.py#L1929-L2127

I'm not sure what the question is about `build`?

**Comment by vdemeester** [comment]
`logging` case and `build` taken into account. I need to make a PR upstream to fix sthg in mergo though.

File: cli/command/stack/kubernetes/deploy.go [hunk 16]
```diff
@@ -16,8 +16,8 @@ import (
 func RunDeploy(dockerCli *KubeCli, opts options.Deploy) error {
 	cmdOut := dockerCli.Out()
 	// Check arguments
-	if opts.Composefile == "" {
-		return errors.Errorf("Please specify a Compose file (with --compose-file).")
+	if len(opts.Composefiles) == 0 {
+		return errors.Errorf("Please specify only one compose file (with --compose-file).")
```
Comment (dnephin) [inline cli/command/stack/kubernetes/deploy.go]: This feature doesn't work with kubernetes deploy?

File: cli/command/stack/swarm/deploy_composefile.go [hunk 17]
```diff
@@ -120,29 +131,32 @@ func propertyWarnings(properties map[string]string) string {
...
-func getConfigDetails(composefile string, stdin io.Reader) (composetypes.ConfigDetails, error) {
+func getConfigDetails(composefiles []string, stdin io.Reader) (composetypes.ConfigDetails, error) {
 	var details composetypes.ConfigDetails
 
-	if composefile == "-" {
+	if len(composefiles) == 0 {
+		return details, errors.New("no composefile(s)")
+	}
+
+	if composefiles[0] == "-" {
```
Comment (dnephin) [inline cli/command/stack/swarm/deploy_composefile.go]: should we error here `if len(composefiles) > 1` ?

File: cli/compose/loader/loader_multiple_test.go [hunk 18]
```diff
@@ -0,0 +1,315 @@
...
+			{Filename: "base.yml", Config: map[string]interface{}{
+				"version": "3.1",
+			}},
+			{Filename: "override.yml", Config: map[string]interface{}{
+				"version": "3.4",
+			}},
+		},
+	}
+	_, err := Load(configDetails)
+	require.Error(t, err)
```
Comment (dnephin) [inline cli/compose/loader/loader_multiple_test.go]: should check the error with `EqualError()`

File: cli/compose/loader/loader_multiple_test.go [hunk 19]
```diff
@@ -0,0 +1,315 @@
+package loader
```
Comment (dnephin) [inline cli/compose/loader/loader_multiple_test.go]: I know loader_test.go is already very long, but I think we need to keep the convention of putting tests into a file that matches the filename of the function being tested.  Otherwise it becomes difficult to find tests and figure out what a test is supposed to be testing. Can we keep this in `loader_test.go`, or maybe put the logging test into `merge_test.go` since that is mostly what it's testing?

File: cli/compose/loader/loader_multiple_test.go [hunk 20]
```diff
@@ -0,0 +1,315 @@
...
+
+func TestLoadLogging(t *testing.T) {
+	loggingCases := []struct {
+		name            string
+		loggingBase     map[string]interface{}
+		loggingOverride map[string]interface{}
+		expected        *types.LoggingConfig
+	}{
+		{
+			name: "options",
```
Comment (dnephin) [inline cli/compose/loader/loader_multiple_test.go]: maybe `"no_override_driver"` ?

File: cli/compose/loader/loader_multiple_test.go [hunk 21]
```diff
@@ -0,0 +1,315 @@
...
+				Networks: map[string]types.NetworkConfig{},
+				Volumes:  map[string]types.VolumeConfig{},
+				Secrets:  map[string]types.SecretConfig{},
+				Configs:  map[string]types.ConfigObjConfig{},
+			}, config)
+		})
+	}
+}
+
+func TestSimpleLoad(t *testing.T) {
```
Comment (dnephin) [inline cli/compose/loader/loader_multiple_test.go]: `TestLoadMultipleConfigs`  so it's clear how this is different from `TestLoad`, and `TestFullExample` (which needs to be renamed to `TestLoadFull`).

File: cli/command/stack/kubernetes/deploy.go (hunk 16 above)
Comment (vdemeester) [reply cli/command/stack/kubernetes/deploy.go]: @dnephin not right now because we parse the file differently. I want to tackle this in a follow-up (i.e. use the same loading/parsing code for both swarm and k8s and thus remove this limitation).

File: cli/command/stack/swarm/deploy_composefile.go (hunk 17 above)
Comment (vdemeester) [reply cli/command/stack/swarm/deploy_composefile.go]: ohh, right probably. How do we handle that on `docker-compose` ?

File: cli/compose/loader/loader_multiple_test.go (hunk 19 above)
Comment (vdemeester) [reply cli/compose/loader/loader_multiple_test.go]: fair point.

File: cli/command/stack/swarm/deploy_composefile.go (hunk 17 above)
Comment (dnephin) [reply cli/command/stack/swarm/deploy_composefile.go]: https://github.com/docker/compose/blob/master/compose/config/config.py#L263-L269

Looks like if it's not `[]string{"-"}` then it tries it as a file now, and would fail later on with a file not found.

**Review by dnephin — APPROVED** [review]
LGTM

**Comment by thaJeztah** [comment]
Some testing (will post more);

file1.yml:

```yaml
version: "3.5"
services:
  web:
    image: nginx:alpine
    ports:
      - "8080:80"
```

file2.yml:

```yaml
version: "3.5"
services:
  web:
    image: nginx:alpine
    ports:
      - target: 80
        published: 8080
        protocol: tcp
        mode: host
```


```
Updating service multi_web (id: l4irmgy0znyncwv1t2ehtbgbd)
failed to update service multi_web: Error response from daemon: rpc error: code = InvalidArgument desc = EndpointSpec: duplicate published ports provided
```

Should the client here detect the duplicate definition? This is what's sent by the CLI:

```son
  "EndpointSpec": {
    "Ports": [
      {
        "Protocol": "tcp",
        "PublishMode": "ingress",
        "PublishedPort": 8080,
        "TargetPort": 80
      },
      {
        "Protocol": "tcp",
        "PublishMode": "host",
        "PublishedPort": 8080,
        "TargetPort": 80
      }
    ]
  },
```

**Comment by thaJeztah** [comment]
Discussing with @vdemeester - there may be corner cases that are not covered yet, and for which we need to define the "correct" (and/or "expected") behaviour.

I think it would be good to make this feature "experimental" at first, and (if experimental cli features are disabled), produce an error if multiple compose files are passed.

**Comment by vdemeester** [comment]
@dnephin @thaJeztah should be ready for another round of review 👼

**Review by dnephin — APPROVED** [review]
LGTM

**Review by thaJeztah — APPROVED** [review]
LGTM 🎉

**Comment by thaJeztah** [comment]
As a follow up, we'll need;

- updates to the completion scripts(?)
- updates for the CLI reference
- possibly other parts of the docs?

**Comment by vlyubin** [comment]
@thaJeztah so what exactly is the CLI syntax? Just `docker stack deploy blah -c file1.yml -c file2.yml`?

Thanks!

**Comment by MetalArend** [comment]
Thanks for implementing this, so glad this finally got to swarm.

I have an edge case that keeps throwign an error. The two files are:

file A:
```
version: '3.7'
volumes:
  app:
    driver: local
    driver_opts:
      o: bind
      type: none
      device: $PWD
```
file B:
```
version: '3.7'
volumes:
  app:
```

If file A is overwritten by file B, everything is fine, but if file B gets overwritten by file A, the following error gets thrown:
```
panic: reflect: reflect.Value.Set using unaddressable value

goroutine 1 [running]:
reflect.flag.mustBeAssignable(0x95)
        /usr/local/go/src/reflect/value.go:234 +0x15c
reflect.Value.Set(0x51b9aa0, 0xc420a449d0, 0x95, 0x51b9aa0, 0xc4206b7b30, 0x15)
        /usr/local/go/src/reflect/value.go:1367 +0x2f
github.com/docker/cli/vendor/github.com/imdario/mergo.deepMerge(0x51b9aa0, 0xc420a449d0, 0x95, 0x51b9aa0, 0xc420a44980, 0x95, 0xc420a1f5b0, 0x2, 0xc420568920, 0x0, ...)
        /go/src/github.com/docker/cli/vendor/github.com/imdario/mergo/merge.go:82 +0xcc0
github.com/docker/cli/vendor/github.com/imdario/mergo.deepMerge(0x52f6540, 0xc420a449b0, 0x99, 0x52f6540, 0xc420a44960, 0x99, 0xc420a1f5b0, 0x1, 0xc420568920, 0x0, ...)
        /go/src/github.com/docker/cli/vendor/github.com/imdario/mergo/merge.go:71 +0x1836
github.com/docker/cli/vendor/github.com/imdario/mergo.deepMerge(0x51b7b20, 0xc420374198, 0x195, 0x51b7b20, 0xc4203741a0, 0x195, 0xc420a1f5b0, 0x0, 0xc420568920, 0x195, ...)
        /go/src/github.com/docker/cli/vendor/github.com/imdario/mergo/merge.go:114 +0x6ec
github.com/docker/cli/vendor/github.com/imdario/mergo._map(0x50f0520, 0xc420374198, 0x50f0520, 0xc4203741a0, 0xc420a1f7b0, 0x1, 0x1, 0x1, 0xc4203741a0)
        /go/src/github.com/docker/cli/vendor/github.com/imdario/mergo/map.go:159 +0x38b
github.com/docker/cli/vendor/github.com/imdario/mergo.Map(0x50f0520, 0xc420374198, 0x50f0520, 0xc4203741a0, 0xc420a1f7b0, 0x1, 0x1, 0x1, 0x0)
        /go/src/github.com/docker/cli/vendor/github.com/imdario/mergo/map.go:132 +0x71
github.com/docker/cli/cli/compose/loader.mergeVolumes(0xc420606690, 0xc4206b7410, 0x8, 0xc420063400, 0x5)
        /go/src/github.com/docker/cli/cli/compose/loader/merge.go:216 +0xd1
github.com/docker/cli/cli/compose/loader.merge(0xc4205f8250, 0x2, 0x2, 0x1, 0x2, 0xc4205f8250)
        /go/src/github.com/docker/cli/cli/compose/loader/merge.go:31 +0x14d
github.com/docker/cli/cli/compose/loader.Load(0xc420601f68, 0x3, 0xc42035cbc0, 0x32, 0xc420552cc0, 0x2, 0x2, 0xc420552cf0, 0x0, 0x0, ...)
        /go/src/github.com/docker/cli/cli/compose/loader/loader.go:111 +0x65f
github.com/docker/cli/cli/command/stack/loader.LoadComposefile(0x5587760, 0xc420167ce0, 0x0, 0x0, 0xc42025b4e0, 0x2, 0x2, 0x7fff5fbffc29, 0x3, 0x53c9fbd, ...)
        /go/src/github.com/docker/cli/cli/command/stack/loader/loader.go:28 +0x18c
github.com/docker/cli/cli/command/stack.newDeployCommand.func1(0xc420566c80, 0xc420369c20, 0x1, 0x6, 0x0, 0x0)
        /go/src/github.com/docker/cli/cli/command/stack/deploy.go:49 +0x23b
github.com/docker/cli/vendor/github.com/spf13/cobra.(*Command).execute(0xc420566c80, 0xc42003a0c0, 0x6, 0x6, 0xc420566c80, 0xc42003a0c0)
        /go/src/github.com/docker/cli/vendor/github.com/spf13/cobra/command.go:762 +0x468
github.com/docker/cli/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc420392f00, 0xc420294330, 0x51389e0, 0xc420294340)
        /go/src/github.com/docker/cli/vendor/github.com/spf13/cobra/command.go:852 +0x30a
github.com/docker/cli/vendor/github.com/spf13/cobra.(*Command).Execute(0xc420392f00, 0xc420392f00, 0x554c4a0)
        /go/src/github.com/docker/cli/vendor/github.com/spf13/cobra/command.go:800 +0x2b
main.main()
        /go/src/github.com/docker/cli/cmd/docker/docker.go:174 +0xd0
```

**Comment by mbrodala** [comment]
Rather ancient but are the merge rules defined formally somewhere? I noticed especially service `volumes` and toplevel `secrets` do not merge well which makes it necessary to fully define them in overrides.

**Comment by thaJeztah** [comment]
@mbrodala could you open a ticket with a minimal example? Note that with Swarm services, "config" and "secret" are swarm objects, unlike docker compose, which doesn't have native support for secrets and configs, and implements them as bind-mounts.

**Comment by mbrodala** [comment]
@thaJeztah Did that now, see https://github.com/docker/cli/issues/7264 for the `secrets` behavior.

I checked the `volumes` behavior once more and don't think there's an issue here after all. I was trying to set `type` and `target` in a base file and only `source` in an override file. But 1) the docs clearly say that the `target` is used as key to determine entries to merge and 2) Compose behaves the same. So this is currently simply not possible.
